boot/freeldr/freeldr/arch/uefi/uefildr.c

```diff
@@ -25,15 +30,65 @@ EfiEntry(
     GlobalImageHandle = ImageHandle;
     GlobalSystemTable = SystemTable;
 
-    BootMain(NULL);
+    /* Debugger pre-initialization */
+    DebugInit(0);
+
+    MachInit("");
+
+    /* UI pre-initialization */
+    if (!UiInitialize(FALSE))
+    {
+        UiMessageBoxCritical("Unable to initialize UI.");
+        goto Quit;
+    }
+
+    /* Initialize memory manager */
+    if (!MmInitializeMemoryManager())
+    {
+        UiMessageBoxCritical("Unable to initialize memory manager.");
+        goto Quit;
+    }
+
+    /* Initialize I/O subsystem */
+    FsInit();
+
+    /* 0x32000 is what UEFI defines, but we can go smaller if we want */
+    BasicStack = (PVOID)((ULONG_PTR)0x32000 + (ULONG_PTR)MmAllocateMemoryWithType(0x32000, LoaderLoadedProgram));
+    _changestack();
+
+Quit:
+    /* If we reach this point, something went wrong before, therefore reboot */
+    Reboot();
 
     UNREACHABLE;
     return 0;
 }
 
+void
+ExecuteLoaderCleanly(PVOID PreviousStack)
+{
+    UefiServiceStack = PreviousStack;
+    if (UefiServiceStack == NULL)
```

Comment:
This would be NULL if PreviousStack was actually NULL I guess. But if in your call you call ExecuteLoaderCleanly by passing a non-NULL parameter, I don't see how this would ever happen.